### PR TITLE
Adding metrics for Encryption

### PR DESCRIPTION
**Pull request by nsivabalan**
1. Adding new set of RouterMetrics for encryption. 
2. Adding new set of encryption metrics for RestRequestMetrics.
3. Fixing a test in replication to test a mix of encrypted and non-encrypted blobs

Coding style applied
Reviewers: Gopal
SLA: 20 mins

### CODE REVIEW

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java [hunk 1]
```diff
@@ -189,27 +189,15 @@ public void handleGet(final RestRequest restRequest, final RestResponseChannel r
...
-                  : frontendMetrics.getReplicasMetrics;
-              securityCallback = new SecurityProcessRequestCallback(restRequest, restResponseChannel);
-              break;
+          if (subresource == SubResource.Replicas) {
+            securityCallback = new SecurityProcessRequestCallback(restRequest, restResponseChannel);
           }
         }
-        restRequest.getMetricsTracker().injectMetrics(requestMetrics);
+        restRequest.getMetricsTracker()
+            .injectMetrics(getRestRequestMetricsForGET(frontendMetrics, subresource, isSsl, false));
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: `..Get` - just like `handleGet()`. Coding style says that three letter words have to be camel case

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java [hunk 2]
```diff
@@ -225,10 +213,12 @@ public void handlePost(RestRequest restRequest, RestResponseChannel restResponse
     long processingStartTime = System.currentTimeMillis();
     long preProcessingTime = 0;
     handlePrechecks(restRequest, restResponseChannel);
-    restRequest.getMetricsTracker().injectMetrics(frontendMetrics.postBlobMetrics);
-    RestRequestMetrics requestMetrics =
-        restRequest.getSSLSession() != null ? frontendMetrics.postBlobSSLMetrics : frontendMetrics.postBlobMetrics;
-    restRequest.getMetricsTracker().injectMetrics(requestMetrics);
+    boolean isSsl = restRequest.getSSLSession() != null;
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: isn't it cleaner the old way (ternary operator)? just delete line 228 (which looks redundant)

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java [hunk 3]
```diff
@@ -135,6 +135,13 @@ public void processResponse(RestRequest restRequest, RestResponseChannel respons
             responseChannel.setStatus(options.getRange() == null ? ResponseStatus.Ok : ResponseStatus.PartialContent);
             responseChannel.setHeader(RestUtils.Headers.LAST_MODIFIED,
                 new Date(blobInfo.getBlobProperties().getCreationTimeInMs()));
+            // inject encryption metrics if applicable
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java]: just do this in the `GetCallback` in `AmbryBlobStorageService`

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java [hunk 4]
```diff
@@ -163,6 +170,12 @@ public void processResponse(RestRequest restRequest, RestResponseChannel respons
                   setAccountAndContainerHeaders(restRequest, responseChannel);
                 }
               }
+              // inject encryption metrics if applicable
+              if (blobInfo.getBlobProperties().isEncrypted()) {
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java]: same here

**Comment by nsivabalan** [comment]
@vgkholla : have updated the patch

File: ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java [hunk 5]
```diff
@@ -392,4 +384,50 @@ public FrontendMetrics(MetricRegistry metricRegistry) {
     putWithAccountAndContainerHeaderRate = metricRegistry.meter(
         MetricRegistry.name(AmbryBlobStorageService.class, "PutWithAccountAndContainerHeaderRate"));
   }
+
+  /**
+   * Class to hold different {@link RestRequestMetrics} for plain text, ssl, with and without encryption
+   */
+  class RestRequestMetricsGroup {
+    private final RestRequestMetrics blobMetrics;
+    private final RestRequestMetrics blobSSLMetrics;
```
Comment (cgtz) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java]: minor: I used the wrong capitalization on these variables. Could we change these to "Ssl" instead of "SSL". That will better match the actual metric name.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java [hunk 6]
```diff
@@ -392,4 +384,50 @@ public FrontendMetrics(MetricRegistry metricRegistry) {
...
+     * Instantiates {@link RestRequestMetricsGroup} for the given requestType and resource
+     * @param requestType refers to the type of request (Head/Get/Post)
+     * @param resource refers to the resource type (Blob/BlobInfo/UserMetadata)
+     */
+    RestRequestMetricsGroup(String requestType, String resource) {
+      blobMetrics = new RestRequestMetrics(AmbryBlobStorageService.class, requestType + resource, metricRegistry);
+      blobSSLMetrics =
+          new RestRequestMetrics(AmbryBlobStorageService.class, requestType + resource + SSL_SUFFIX, metricRegistry);
+      encryptedBlobMetrics =
+          new RestRequestMetrics(AmbryBlobStorageService.class, requestType + ENCRYPTED + resource, metricRegistry);
```
Comment (cgtz) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java]: Would you rather have the Encrypted in the middle of the name, or at the end? Depends on how you think you will tend to search for these metrics

File: ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java [hunk 7]
```diff
@@ -392,4 +384,50 @@ public FrontendMetrics(MetricRegistry metricRegistry) {
...
+              metricRegistry);
+    }
+
+    /**
+     * Fetches the appropriate {@link RestRequestMetrics} based on the params
+     * @param isSsl {@code true} if the request is over ssl. {@code false} otherwise
+     * @param encrypted {@code true} if the blob is encrypted. {@code false} otherwise
+     * @return the appropriate {@link RestRequestMetrics} based on the params
+     */
+    RestRequestMetrics getRestRequestMetrics(boolean isSsl, boolean encrypted) {
```
Comment (cgtz) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java]: I think in other parts of the code, I used `sslUsed` as a variable name.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java (hunk 6 above)
Comment (nsivabalan) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java]: Moved encryption towards the end.

**Comment by nsivabalan** [comment]
@cgtz @vgkholla : can we get this merged if things look good.

**Comment by vgkholla** [comment]
I haven't reviewed it completely

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java [hunk 8]
```diff
@@ -402,10 +402,23 @@ private void checkAndMaybeComplete() {
...
+        boolean encrypted = operationResult != null && operationResult.getBlobResult != null
+            && operationResult.getBlobResult.getBlobInfo() != null && operationResult.getBlobResult.getBlobInfo()
+            .getBlobProperties()
+            .isEncrypted();
         operationResult = null;
-        routerMetrics.onGetBlobError(e, options);
+        routerMetrics.onGetBlobError(e, options, encrypted);
+      }
+      long operationLatencyMs = time.milliseconds() - submissionTimeMs;
+      routerMetrics.getBlobInfoOperationLatencyMs.update(operationLatencyMs);
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: if you update the main one anyway, wouldn't it be against our goal of trying to separate the encrypted from the not encrypted metrics?

Comment (vgkholla) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: Same comment in all other ops - want to understand what the intent here since in the frontend you separate them but in the router it looks like one metric contains both and another has just encrypted.

Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: yes. I did start the patch with that in mind. But for gets, we may not know if the blob is encrypted or not until we get the response. So, I chose this path. regular metrics will include both non-encrypted and encrypted blobs. encryption metrics will be emitted in addition for encrypted blobs.

Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: oh yeah. I missed that part. Once we have a consensus, will fix the diff between frontend metrics and router metrics.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java [hunk 9]
```diff
@@ -225,10 +213,9 @@ public void handlePost(RestRequest restRequest, RestResponseChannel restResponse
     long processingStartTime = System.currentTimeMillis();
     long preProcessingTime = 0;
     handlePrechecks(restRequest, restResponseChannel);
-    restRequest.getMetricsTracker().injectMetrics(frontendMetrics.postBlobMetrics);
-    RestRequestMetrics requestMetrics =
-        restRequest.getSSLSession() != null ? frontendMetrics.postBlobSSLMetrics : frontendMetrics.postBlobMetrics;
-    restRequest.getMetricsTracker().injectMetrics(requestMetrics);
+    boolean sslUsed = restRequest.getSSLSession() != null;
+    RestRequestMetrics metrics = frontendMetrics.postRequestMetricsGroup.getRestRequestMetrics(sslUsed, false);
+    restRequest.getMetricsTracker().injectMetrics(metrics);
```
Comment (cgtz) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: So we set up metrics up here and maybe overwrite it later, as opposed to just setting it up on line 232 so that the restRequest has a metrics instance when `securityService.preProcessRequest()` is called?

File: ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java [hunk 10]
```diff
@@ -575,7 +575,7 @@ public void replicaTokenTest() throws InterruptedException {
       BlobId id = new BlobId(blobIdVersion, BlobId.BlobIdType.NATIVE, ClusterMapUtils.UNKNOWN_DATACENTER_ID, accountId,
           containerId, partitionId);
       ids.add(id);
-      Pair<ByteBuffer, MessageInfo> putMsgInfo = getPutMessage(id, accountId, containerId);
+      Pair<ByteBuffer, MessageInfo> putMsgInfo = getPutMessage(id, accountId, containerId, i % 2 == 0 ? true : false);
```
Comment (cgtz) [inline ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java]: can just use `i % 2 == 0` instead of `i % 2 == 0 ? true : false`

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java [hunk 11]
```diff
@@ -1692,14 +1692,19 @@ private static void checkBlobContent(MockClusterMap clusterMap, BlobId blobId, B
     }
   }
 
-  private static Properties getRouterProps(String routerDatacenter) {
+  /**
+   *
+   * @param routerDatacenter
+   * @return
+   */
+  private static Properties getRouterProps(String routerDatacenter, boolean enableCrossDC) {
```
Comment (cgtz) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: where is `enableCrossDc` false? Also, the javadoc is incomplete.

**Comment by nsivabalan** [comment]
@cgtz : will update the patch(and will address your comments too) once https://github.com/linkedin/ambry/pull/822 is merged.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java [hunk 12]
```diff
@@ -189,27 +189,15 @@ public void handleGet(final RestRequest restRequest, final RestResponseChannel r
         GetCallback routerCallback = new GetCallback(restRequest, restResponseChannel, subresource, options);
         SecurityProcessRequestCallback securityCallback =
             new SecurityProcessRequestCallback(restRequest, restResponseChannel, routerCallback);
-        RestRequestMetrics requestMetrics =
-            restRequest.getSSLSession() != null ? frontendMetrics.getBlobSSLMetrics : frontendMetrics.getBlobMetrics;
+        boolean isSsl = restRequest.getSSLSession() != null;
         if (subresource != null) {
```
[the remark this answers is not in the document]
Comment (nsivabalan) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: I would leave it this way. Doesn't make much diff.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java (hunk 9 above)
Comment (nsivabalan) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: yes. you are right.

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java (hunk 11 above)
Comment (nsivabalan) [reply ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: will fix it

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java [hunk 13]
```diff
@@ -189,30 +190,23 @@ public void handleGet(final RestRequest restRequest, final RestResponseChannel r
...
-              requestMetrics = restRequest.getSSLSession() != null ? frontendMetrics.getReplicasSSLMetrics
-                  : frontendMetrics.getReplicasMetrics;
-              securityCallback = new SecurityProcessRequestCallback(restRequest, restResponseChannel);
-              break;
+          if (subresource == SubResource.Replicas) {
+            securityCallback = new SecurityProcessRequestCallback(restRequest, restResponseChannel);
           }
         }
-        restRequest.getMetricsTracker().injectMetrics(requestMetrics);
+        boolean isEncrypted = new BlobId(operationOrBlobId, clusterMap).isEncrypted();
```
[the remark this answers is not in the document]
Comment (nsivabalan) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: thanks for pointing it out. I will create BlobId the same way as you suggested, but will leave injectTargetAccountAndContainerFromBlobId() as is. Don't wanna change it to take in BlobId. If not, have to change InBoundIdConverterCallback() impl or api which is uncalled for considering this patch.

[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: Not sure I understand. You can have to create the `BlobId` in `InboundIdConverterCallback` - so what's the extra work required to make `AccountAndContainerInjector#injectTargetAccountAndContainerFromBlobId()` take in a `BlobId`? You can't generate a `BlobId` at this point anyway

[the remark this answers is not in the document]
Comment (nsivabalan) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: sorry, why do I have to create BlobId in InboundIdConverterCallback?

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java [hunk 14]
```diff
@@ -189,30 +190,23 @@ public void handleGet(final RestRequest restRequest, final RestResponseChannel r
...
+        restRequest.getMetricsTracker()
+            .injectMetrics(
+                getRestRequestMetricsForGet(frontendMetrics, subresource, restRequest.getSSLSession() != null,
+                    isEncrypted));
         securityService.processRequest(restRequest, securityCallback);
       }
       preProcessingTime = System.currentTimeMillis() - processingStartTime;
+    } catch (IllegalArgumentException | IOException iaOrioException) {
```
[the remark this answers is not in the document]
Comment (nsivabalan) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: have fixed it.

**Comment by nsivabalan** [comment]
@vgkholla : regarding your comments on line 218 in AmbryBlobStorageService ( "There is no necessity to do that. Metrics can be injected at any point and they will be tracked against the correct instance."). If at all we encounter any exception before we have the BlobProperties in hand, (for instance in injectAccountAndContainerForPostRequest can throw exception), we may not have the metrics instance attached to RestRequest and hence I had to do this. Let me know if my understanding is wrong.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: As mentioned in my original comment it is because it is possible that the ID at this point is not a valid Ambry ID since the ID converter has not yet run

**Comment by nsivabalan** [comment]
@vgkholla : have updated after addressing your comments.

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 15]
```diff
@@ -891,6 +907,7 @@ private void encryptChunk() {
     void onFillComplete(boolean updateMetric) {
       buf.flip();
       chunkBlobSize = buf.remaining();
+      chunkFillCompleteAtMs = time.milliseconds();
```
Comment (cgtz) [inline ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: where is this used?

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java [hunk 16]
```diff
@@ -492,12 +511,26 @@ public void onCompletion(String result, Exception exception) {
         try {
           RestMethod restMethod = restRequest.getRestMethod();
           logger.trace("Handling {} of {}", restMethod, result);
-          accountAndContainerInjector.injectTargetAccountAndContainerFromBlobId(result, restRequest);
+          BlobId blobId;
+          try {
```
Comment (cgtz) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: Since this try-catch code is used in multiple places, you can consider making a static util method

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java (hunk 15 above)
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: check line 318 within onChunkOperationComplete()

**Comment by nsivabalan** [comment]
@cgtz @vgkholla : addressed comments

**Comment by vgkholla** [comment]
@nsivabalan did you push all your changes? I still see `getBlobIdFromString()` in `BlobId`

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 17]
```diff
@@ -498,6 +500,21 @@ public static boolean isCrafted(String idStr) throws IOException {
...
   }
 
+  /**
+   * Fetches {@link BlobId} from the given string representation of BlobId
+   * @param blobIdStr string representation of BlobId
+   * @param clusterMap {@link ClusterMap} instance to use
+   * @return the {@link BlobId} thus generated
+   * @throws RestServiceException
+   */
+  public static BlobId getBlobIdFromString(String blobIdStr, ClusterMap clusterMap) throws RestServiceException {
```
Comment (cgtz) [inline ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: Make sure to clean up this method

Comment (nsivabalan) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: oops. missed to remove it. Have removed it now.

**Comment by nsivabalan** [comment]
@vgkholla : sorry to have missed to remove it. you can review it now

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AccountAndContainerInjector.java [hunk 18]
```diff
@@ -88,40 +88,33 @@ void injectAccountAndContainerForPostRequest(RestRequest restRequest) throws Res
...
    * @param restRequest The rest request to insert the target {@link Account} and {@link Container}.
    * @throws RestServiceException if 1) either {@link Account} or {@link Container} could not be found; or 2)
    *                              either {@link Account} or {@link Container} were explicitly specified as
    *                              {@link Account#UNKNOWN_ACCOUNT} or {@link Container#UNKNOWN_CONTAINER}.
    */
-  void injectTargetAccountAndContainerFromBlobId(String blobIdStr, RestRequest restRequest)
-      throws RestServiceException {
-    BlobId blobId;
-    try {
-      blobId = new BlobId(blobIdStr, clusterMap);
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AccountAndContainerInjector.java]: there is a `ClusterMap` instance in this class. It must be unused now?

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java [hunk 19]
```diff
@@ -492,12 +511,21 @@ public void onCompletion(String result, Exception exception) {
         try {
           RestMethod restMethod = restRequest.getRestMethod();
           logger.trace("Handling {} of {}", restMethod, result);
-          accountAndContainerInjector.injectTargetAccountAndContainerFromBlobId(result, restRequest);
+          BlobId blobId = FrontendUtils.getBlobIdFromString(result, clusterMap);
+          boolean isEncrypted = blobId.isEncrypted();
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java]: minor: this seems unnecessary

File: ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java [hunk 20]
```diff
@@ -392,4 +386,50 @@ public FrontendMetrics(MetricRegistry metricRegistry) {
     putWithAccountAndContainerHeaderRate = metricRegistry.meter(
         MetricRegistry.name(AmbryBlobStorageService.class, "PutWithAccountAndContainerHeaderRate"));
   }
+
+  /**
+   * Class to hold different {@link RestRequestMetrics} for plain text, ssl, with and without encryption
+   */
+  class RestRequestMetricsGroup {
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java]: `private static`

File: ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java [hunk 21]
```diff
@@ -392,4 +386,50 @@ public FrontendMetrics(MetricRegistry metricRegistry) {
...
+    private final RestRequestMetrics blobEncryptedMetrics;
+    private final RestRequestMetrics blobSslEncryptedMetrics;
+
+    /**
+     * Instantiates {@link RestRequestMetricsGroup} for the given requestType and resource
+     * @param requestType refers to the type of request (Head/Get/Post)
+     * @param resource refers to the resource type (Blob/BlobInfo/UserMetadata)
+     */
+    RestRequestMetricsGroup(String requestType, String resource) {
+      blobMetrics = new RestRequestMetrics(AmbryBlobStorageService.class, requestType + resource, metricRegistry);
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java]: very minor: is the "blob" prefix on all of these variables required?

File: ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendUtils.java [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+ * Common utility functions that will be used across frontend package
+ */
+class FrontendUtils {
+
+  /**
+   * Fetches {@link BlobId} from the given string representation of BlobId
+   * @param blobIdStr string representation of BlobId
+   * @param clusterMap {@link ClusterMap} instance to use
+   * @return the {@link BlobId} thus generated
+   * @throws RestServiceException
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendUtils.java]: document when

File: ambry-frontend/src/test/java/com.github.ambry.frontend/FrontendUtilsTest.java [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+   */
+  private void assertBlobIdFieldValues(BlobId blobId, BlobId.BlobIdType type, byte datacenterId, short accountId,
+      short containerId, PartitionId partitionId, boolean isEncrypted) throws IOException {
+    assertEquals("Wrong partition id in blobId: " + blobId, partitionId, blobId.getPartition());
+    assertEquals("Wrong type in blobId: " + blobId, type, blobId.getType());
+    assertEquals("Wrong datacenter id in blobId: " + blobId, datacenterId, blobId.getDatacenterId());
+    assertEquals("Wrong account id in blobId: " + blobId, accountId, blobId.getAccountId());
+    assertEquals("Wrong container id in blobId: " + blobId, containerId, blobId.getContainerId());
+    assertEquals("Wrong isEncrypted value in blobId: " + blobId, isEncrypted, blobId.isEncrypted());
+    Pair<Short, Short> accountAndContainer = BlobId.getAccountAndContainerIds(blobId.getID());
```
Comment (vgkholla) [inline ambry-frontend/src/test/java/com.github.ambry.frontend/FrontendUtilsTest.java]: this part is not required. This should be tested in `BlobId` (if it is not, just transfer this piece of code there)

File: ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java [hunk 24]
```diff
@@ -322,11 +322,12 @@ public void replicaThreadTest() throws Exception {
```
Comment (vgkholla) [inline ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java]: need to chat about whether this guarantees coverage of all the cases we want it to (if it does not, lets make a separate PR for that since this is unrelated to metrics)

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java [hunk 25]
```diff
@@ -402,10 +402,16 @@ private void checkAndMaybeComplete() {
         routerMetrics.operationFailureWithUnsetExceptionCount.inc();
       }
       if (e != null) {
+        // best effort to update metrics if result is available
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: does this comment belong here? (or even needed now that the blob ID has the details on encrypted or not)

File: ambry-router/src/main/java/com.github.ambry.router/GetManager.java [hunk 26]
```diff
@@ -128,7 +128,7 @@ void submitGetBlobOperation(String blobId, GetBlobOptionsInternal options, Callb
       }
       getOperations.add(getOperation);
     } catch (RouterException e) {
-      routerMetrics.onGetBlobError(e, options);
+      routerMetrics.onGetBlobError(e, options, false);
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetManager.java]: it seems like the right thing to do is to convert the string to `BlobId` early in the router stack.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 27]
```diff
@@ -144,17 +146,31 @@ private OperationController getOperationController() {
       throw new IllegalArgumentException("blobId or options must not be null");
     }
     currentOperationsCount.incrementAndGet();
+    final FutureResult<GetBlobResult> futureResult = new FutureResult<>();
+    GetBlobOptionsInternal internalOptions = new GetBlobOptionsInternal(options, false, routerMetrics.ageAtGet);
+    boolean isEncrypted = false;
+    try {
+      isEncrypted = new BlobId(blobId, clusterMap).isEncrypted();
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: 1) There is a function for this `RouterUtils.getBlobIdFromString()`
2) It looks like you convert to `BlobId` early in the stack, why not pass it down? I know it will cause a few constructor changes but before now, we didn't need a string -> `BlobId` conversion so early. Now we do.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 28]
```diff
@@ -144,17 +146,31 @@ private OperationController getOperationController() {
       throw new IllegalArgumentException("blobId or options must not be null");
     }
     currentOperationsCount.incrementAndGet();
+    final FutureResult<GetBlobResult> futureResult = new FutureResult<>();
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: why was the location of this line and the next moved?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 29]
```diff
@@ -144,17 +146,31 @@ private OperationController getOperationController() {
...
+    GetBlobOptionsInternal internalOptions = new GetBlobOptionsInternal(options, false, routerMetrics.ageAtGet);
+    boolean isEncrypted = false;
+    try {
+      isEncrypted = new BlobId(blobId, clusterMap).isEncrypted();
+    } catch (IOException e) {
+      RouterException routerException =
+          new RouterException("Exception thrown during construction of BlobId ", e, RouterErrorCode.InvalidBlobId);
+      completeGetBlobOperation(routerException, internalOptions, futureResult, callback);
+    }
+    Meter blobInfoOperationRate =
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: it looks like this code has enough control structures to be put into its own function for readability? something like `trackRate(options, isEncrypted)`

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java [hunk 30]
```diff
@@ -315,6 +333,21 @@ private void initiateChunkDeletesIfAny(final String blobId, final String service
...
+   * {@code {@link RouterException}}
+   * @param routerException {@link RouterException} to be set in the callback and future result
+   * @param internalOptions instance of {@link GetBlobOptionsInternal} to use
+   * @param futureResult the {@link FutureResult} that needs to be set.
+   * @param callback that {@link Callback} that needs to be invoked. Can be null.
+   */
+  private void completeGetBlobOperation(RouterException routerException, GetBlobOptionsInternal internalOptions,
+      FutureResult<GetBlobResult> futureResult, Callback<GetBlobResult> callback) {
+    routerMetrics.operationDequeuingRate.mark();
+    routerMetrics.onGetBlobError(routerException, internalOptions, false);
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: it seems like sometimes you _do_ have access to whether the blob is encrypted or not when you call this function, no?

File: ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java (hunk 20 above)
Comment (nsivabalan) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java]: it uses metricRegistry and hence can't be static. Will add private.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java (hunk 21 above)
Comment (nsivabalan) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java]: I thought about that too. Then, non encryted and non ssl can't be named just "metrics" and might have to go with restReqeustMetrics and inturn will bloat every other metrics variable. Hence went with this. Let me know what is your take here.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 28 above)
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: if BlobId creation throws an exception, we need an handle to futureResult and GetBlobOptionsInternal and hence

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 29 above)
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: Yes, will do move it into separate method

File: ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java (hunk 20 above)
Comment (nsivabalan) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/FrontendMetrics.java]: Sorry. this is being access by AmbryBlobStorageService and hence can't make it private too

File: ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java (hunk 24 above)
Comment (nsivabalan) [reply ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java]: sure. lets take it outside this patch.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java (hunk 27 above)
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java]: have fixed the 1st comment. Will take the 2nd as a separate commit

File: ambry-router/src/main/java/com.github.ambry.router/GetManager.java (hunk 26 above)
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/GetManager.java]: sure. will take it up as a separate commit

**Comment by nsivabalan** [comment]
@vgkholla : updated the patch with fixing the BlobId wiring.

**Comment by vgkholla** [comment]
@nsivabalan thanks for making this change ! Some minor comments. LGTM

**Comment by nsivabalan** [comment]
@vgkholla : I couldn't find one of your comment. "minor: the =null is not required I think?". can you tell me where is this ? line number. I guess its in NonBlockingRouter

**Comment by nsivabalan** [comment]
@vgkholla : have updated the patch

**Comment by vgkholla** [comment]
LGTM. Will merge once the travis build finishes
